[Quant] Add fused linear-leaky_relu op for onednn backend

Xia-Weiwen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89188
* #88923
* #88879
* #88668
* #88665
* #88661
* __->__ #88478

**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `linear-leaky_relu` op for `onednn` backend, which will be used for int8 inference with `onednn` backend. Cannot call this op with other quantization backends otherwise an error is thrown.

**Test Plan**
python test_quantization.py TestQuantizedLinear

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @leslie-fang-intel @VitalyFedyunin @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

Xia-Weiwen committed:
Add quantized linear-leaky_relu fusion for onednn backend

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "Add quantized linear-leaky_relu fusion for onednn backend"




cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen commented:
This is part of replacement of https://github.com/pytorch/pytorch/pull/76424 which is too big to land.

Xia-Weiwen committed:
Update on "Add quantized linear-leaky_relu fusion for onednn backend"




cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused linear-leaky_relu op for onednn backend"




cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused linear-leaky_relu op for onednn backend"




cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -3663,14 +3656,34 @@ def test_qlinear(self, batch_size, input_channels, output_channels, use_bias,
 
+    """Tests the correctness of the quantized linear and linear_relu op."""
+    @override_qengines
+    def test_qlinear(self):
```
looks like this is qlinear + qlinear relu? I feel we can split this to two tests? we would like to make each test a unit test I think

Xia-Weiwen replied:
Ok I will do that.

Xia-Weiwen replied:
I have split it to two separate tests.

jerryzh168 reviewed (approved):
lg, please split the test

jerryzh168 commented:
Also for future PRs, please add a "Summary" that talks about the motivation and context for the change and a "Test Plan" section that talks about how to run the test, e.g.: https://github.com/pytorch/pytorch/pull/88620

Xia-Weiwen commented:
> Also for this one and future PRs, please add a "Summary" that talks about the motivation and context for the change and a "Test Plan" section that talks about how to run the test, e.g.: #88620

Sure. I will add description later.

Xia-Weiwen committed:
Update on "[Quant] Add fused linear-leaky_relu op for onednn backend"

**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `linear-leaky_relu` op for `onednn` backend, which will be used for int8 inference with `onednn` backend. Cannot call this op with other quantization backends otherwise an error is thrown.

**Test Plan**
python test_quantization.py TestQuantizedLinear




cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen commented:
> > Also for this one and future PRs, please add a "Summary" that talks about the motivation and context for the change and a "Test Plan" section that talks about how to run the test, e.g.: #88620
> 
> Sure. I will add description later.

I have added summary and test plan.

Xia-Weiwen committed:
Update on "[Quant] Add fused linear-leaky_relu op for onednn backend"

**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `linear-leaky_relu` op for `onednn` backend, which will be used for int8 inference with `onednn` backend. Cannot call this op with other quantization backends otherwise an error is thrown.

**Test Plan**
python test_quantization.py TestQuantizedLinear




cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 reviewed (approved):
looks good to me

jerryzh168 commented:
btw the Summary and Test Plan should appear after the stack..

Xia-Weiwen committed:
Update on "[Quant] Add fused linear-leaky_relu op for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `linear-leaky_relu` op for `onednn` backend, which will be used for int8 inference with `onednn` backend. Cannot call this op with other quantization backends otherwise an error is thrown.

**Test Plan**
python test_quantization.py TestQuantizedLinear

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen commented:
Hi @jerryzh168.  The separated `test_qlinear` and `test_qlinear_relu` failed but they look good when running locally. I also created a draft PR which only separated the two tests without other changes, just for test. But they also failed.
Do you have any ideas why?

Xia-Weiwen committed:
Update on "[Quant] Add fused linear-leaky_relu op for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `linear-leaky_relu` op for `onednn` backend, which will be used for int8 inference with `onednn` backend. Cannot call this op with other quantization backends otherwise an error is thrown.

**Test Plan**
python test_quantization.py TestQuantizedLinear

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused linear-leaky_relu op for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `linear-leaky_relu` op for `onednn` backend, which will be used for int8 inference with `onednn` backend. Cannot call this op with other quantization backends otherwise an error is thrown.

**Test Plan**
python test_quantization.py TestQuantizedLinear

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused linear-leaky_relu op for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `linear-leaky_relu` op for `onednn` backend, which will be used for int8 inference with `onednn` backend. Cannot call this op with other quantization backends otherwise an error is thrown.

**Test Plan**
python test_quantization.py TestQuantizedLinear

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen commented:
> Hi @jerryzh168. The separated `test_qlinear` and `test_qlinear_relu` failed so merge failed. But they look good when running locally. I also created a draft PR which only separated the two tests without other changes, just for test (#89678) But they also failed. Do you have any ideas why?

Hi, @jerryzh168. It was due to some weird rounding errors, I think. `Y_scale = 125.1234` caused the issue in fbgemm. So, I changed the scale value to `12.34`. Now all checks have passed. How does that sound to you?

Xia-Weiwen committed:
Update on "[Quant] Add fused linear-leaky_relu op for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `linear-leaky_relu` op for `onednn` backend, which will be used for int8 inference with `onednn` backend. Cannot call this op with other quantization backends otherwise an error is thrown.

**Test Plan**
python test_quantization.py TestQuantizedLinear

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel VitalyFedyunin mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/library.cpp:
```diff
@@ -152,6 +152,7 @@ TORCH_LIBRARY(quantized, m) {
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::linear_relu_dynamic(Tensor X, __torch__.torch.classes.quantized.LinearPackedParamsBase W_prepack, bool reduce_range=False) -> Tensor Y"));
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::linear_dynamic_fp16(Tensor X, __torch__.torch.classes.quantized.LinearPackedParamsBase W_prepack) -> Tensor Y"));
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::linear_relu_dynamic_fp16(Tensor X, __torch__.torch.classes.quantized.LinearPackedParamsBase W_prepack) -> Tensor Y"));
+  m.def(TORCH_SELECTIVE_SCHEMA("quantized::linear_leaky_relu(Tensor X, __torch__.torch.classes.quantized.LinearPackedParamsBase W_prepack, float Y_scale_i, int Y_zero_point_i, float negative_slope) -> Tensor Y"));
```
nit: what does `i` in `Y_scale_i` mean?

Xia-Weiwen replied:
> nit: what does `i` in `Y_scale_i` mean?

Good question. I don't know. I just copied the arg name from above to look aligned https://github.com/pytorch/pytorch/blob/55b10e6b1d52aeae536129d152b1d057f159bfc0/aten/src/ATen/native/quantized/library.cpp#L150
Shall I keep it or remove it?

jerryzh168 commented:
yeah LGTM

jerryzh168 commented:
> 
> 
> It has been good before we separated `test_qlinear` and `test_qlinear_relu` but the issue occurred after the separation.
> To avoid the issue and keep the test cases reasonable, I changed the scale value to `12.34`. Now all checks have passed. How does that sound to you?

what is the error? is it a numerical error or something else?

Xia-Weiwen commented:
> > It has been good before we separated `test_qlinear` and `test_qlinear_relu` but the issue occurred after the separation.
> > To avoid the issue and keep the test cases reasonable, I changed the scale value to `12.34`. Now all checks have passed. How does that sound to you?
> 
> what is the error? is it a numerical error or something else?

Yes, it's a numerical error of fbgemm.

Xia-Weiwen commented:
@pytorchbot merge
